test/test_detailed_hydro.jl

```diff
@@ -0,0 +1,194 @@
+    inflow = [hydro_reservoir.vol_inflow[t] for t in 𝒯]
+    @test level_Δ == inflow - discharge
+end
+
+@testset "Test hydro reservoir hard MinConstraint and MaxConstraint" begin
+    case, model = build_case()
+    optimizer = optimizer_with_attributes(HiGHS.Optimizer, MOI.Silent() => true)
+
+    𝒯 = case[:T]
+    hydro_reservoir = case[:nodes][1]
+    hydro_gate = case[:nodes][2]
+    Water = case[:products][3]
+    max_profile = [1, 0.8, 0.8, 1]
+    push!(hydro_reservoir.data,
+        MaxConstraint(
+            Symbol(),
+            OperationalProfile(max_profile), # value
+            FixedProfile(true),                 # flag
+            FixedProfile(Inf),                  # penalty
+        )
+    )
+    min_profile = [1, 0, 0, 0]
+    push!(hydro_reservoir.data,
+        MinConstraint(
+            Symbol(),
+            OperationalProfile(min_profile), # value
+            FixedProfile(true),                 # flag
+            FixedProfile(Inf),                  # penalty
+        )
+    )
+    m = EMB.run_model(case, model, optimizer)
+    for sp in strategic_periods(𝒯)
+        rsv_vol = value.([m[:stor_level][hydro_reservoir, t] for t in sp])
+        min_vol = [hydro_reservoir.vol.capacity[t] for t in sp] .* min_profile
+        max_vol = [hydro_reservoir.vol.capacity[t] for t in sp] .* max_profile
+        @test min_vol ≤ rsv_vol ≤ max_vol
+    end
+end
+
+@testset "Test hydro reservoir hard MaxConstraint penalty cost" begin
+    case, model = build_case()
+    optimizer = optimizer_with_attributes(HiGHS.Optimizer, MOI.Silent() => true)
+
+    𝒯 = case[:T]
+    hydro_reservoir = case[:nodes][1]
+    hydro_gate = case[:nodes][2]
+    Water = case[:products][3]
+
+    # Verify reservoir minimum/maximum hard constraint
+    max_profile = [1, 0, 0.8, 1]
+    penalty_cost = 57
+    push!(hydro_reservoir.data,
+        MaxConstraint(
+            Symbol(),
+            OperationalProfile(max_profile), # value
+            OperationalProfile([false, true, false, false]), # flag
+            FixedProfile(penalty_cost),                      # penalty
+        )
+    )
+    min_profile = [1, 0, 0, 0]
+    push!(hydro_reservoir.data,
+        MinConstraint(
+            Symbol(),
+            OperationalProfile(min_profile), # value
+            FixedProfile(true),              # flag
+            FixedProfile(Inf),               # penalty
+        )
+    )
+    m = EMB.run_model(case, model, optimizer)
+    max_vol_violation_cost = 0
+    for sp in strategic_periods(𝒯)
+        rsv_vol = value.([m[:stor_level][hydro_reservoir, t] for t in sp])
+        min_vol = [hydro_reservoir.vol.capacity[t] for t in sp] .* min_profile
+        max_vol = [hydro_reservoir.vol.capacity[t] for t in sp] .* max_profile
+        max_vol_violation = max.(rsv_vol - max_vol, 0)
+        max_vol_violation_cost += sum(max_vol_violation * penalty_cost)
+    end
+    @test objective_value(m) + max_vol_violation_cost == 0
+end
+
```

Comment:
Endres produksjonsprofilen når min/maks-restriksjonene tas med i disse testene? Altså, er det incentiver for å ligge høyere/lavere enn grensene uten restriksjon?

Reply:
Will update, but will squash these changes into our dev branch first